In [1]:
%matplotlib inline

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

plt.style.use('seaborn') # Make the graphs a bit prettier

plt.rcParams['figure.figsize'] = (15, 3)
plt.rcParams['font.family'] = 'sans-serif'

We saw earlier that pandas is really good at dealing with dates. It is also amazing with strings! We're going to go back to our weather data from Chapter 5, here.

In [2]:
weather_2012 = pd.read_csv('../data/weather_2012.csv', parse_dates=True, index_col='Date/Time')
weather_2012[:5]

Temp (C)  Dew Point Temp (C)  Rel Hum (%)  \
Date/Time                                                        
2012-01-01 00:00:00      -1.8                -3.9           86   
2012-01-01 01:00:00      -1.8                -3.7           87   
2012-01-01 02:00:00      -1.8                -3.4           89   
2012-01-01 03:00:00      -1.5                -3.2           88   
2012-01-01 04:00:00      -1.5                -3.3           88   

                     Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)  \
Date/Time                                                                
2012-01-01 00:00:00                4              8.0           101.24   
2012-01-01 01:00:00                4              8.0           101.24   
2012-01-01 02:00:00                7              4.0           101.26   
2012-01-01 03:00:00                6              4.0           101.27   
2012-01-01 04:00:00                7              4.8           101.23   

                                  Weather  
Date/Time                                  
2012-01-01 00:00:00                   Fog  
2012-01-01 01:00:00                   Fog  
2012-01-01 02:00:00  Freezing Drizzle,Fog  
2012-01-01 03:00:00  Freezing Drizzle,Fog  
2012-01-01 04:00:00                   Fog

# 6.1 Операции со строками

You'll see that the 'Weather' column has a text description of the weather that was going on each hour. We'll assume it's snowing if the text description contains "Snow".

In [3]:
weather_description = weather_2012['Weather']
is_snowing = weather_description.str.contains('Snow')

In [4]:
# Not super useful
is_snowing[:5]

Date/Time
2012-01-01 00:00:00    False
2012-01-01 01:00:00    False
2012-01-01 02:00:00    False
2012-01-01 03:00:00    False
2012-01-01 04:00:00    False
Name: Weather, dtype: bool

In [11]:
# More useful!
np.unique(is_snowing.index.month[is_snowing])

array([ 1,  2,  3,  4, 11, 12], dtype=int32)

# 6.2 Используем resampling

If we wanted the median temperature each month, we could use the `resample()` method like this:

In [17]:
temperature = weather_2012['Temp (C)'].resample('M').apply(np.median)
temperature

Date/Time
2012-01-31    -7.05
2012-02-29    -4.10
2012-03-31     2.60
2012-04-30     6.30
2012-05-31    16.05
2012-06-30    19.60
2012-07-31    22.90
2012-08-31    22.20
2012-09-30    16.10
2012-10-31    11.30
2012-11-30     1.05
2012-12-31    -2.85
Freq: M, Name: Temp (C), dtype: float64

Unsurprisingly, July and August are the warmest.

So we can think of snowiness as being a bunch of 1s and 0s instead of `True`s and `False`s:

In [18]:
is_snowing.astype(float)[:10]

Date/Time
2012-01-01 00:00:00    0.0
2012-01-01 01:00:00    0.0
2012-01-01 02:00:00    0.0
2012-01-01 03:00:00    0.0
2012-01-01 04:00:00    0.0
2012-01-01 05:00:00    0.0
2012-01-01 06:00:00    0.0
2012-01-01 07:00:00    0.0
2012-01-01 08:00:00    0.0
2012-01-01 09:00:00    0.0
Name: Weather, dtype: float64

and then use `resample` to find the percentage of time it was snowing each month

In [19]:
snowiness = is_snowing.astype(float).resample('M').apply(np.mean)
snowiness

Date/Time
2012-01-31    0.240591
2012-02-29    0.162356
2012-03-31    0.087366
2012-04-30    0.015278
2012-05-31    0.000000
2012-06-30    0.000000
2012-07-31    0.000000
2012-08-31    0.000000
2012-09-30    0.000000
2012-10-31    0.000000
2012-11-30    0.038889
2012-12-31    0.251344
Freq: M, Name: Weather, dtype: float64

# 6.3 Помещаем все в одну таблицу

In [20]:
# Name the columns
temperature.name = "Temperature"
snowiness.name = "Snowiness"

We'll use `concat` again to combine the two statistics into a single dataframe.

In [21]:
stats = pd.concat([temperature, snowiness], axis=1)
stats

Temperature  Snowiness
Date/Time                         
2012-01-31        -7.05   0.240591
2012-02-29        -4.10   0.162356
2012-03-31         2.60   0.087366
2012-04-30         6.30   0.015278
2012-05-31        16.05   0.000000
2012-06-30        19.60   0.000000
2012-07-31        22.90   0.000000
2012-08-31        22.20   0.000000
2012-09-30        16.10   0.000000
2012-10-31        11.30   0.000000
2012-11-30         1.05   0.038889
2012-12-31        -2.85   0.251344

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  